keps/sig-cli/3517-kubectl-client-side-apply-migration/README.md

```diff
@@ -0,0 +1,1113 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Adds a migration path to use server-side-apply from usage of client-side-apply.
+
+## Motivation
+
+The current documentation for Kubernetes Server-Side-Apply has [instructions for
+migrating](https://kubernetes.io/docs/reference/using-api/server-side-apply/#upgrading-from-client-side-apply-to-server-side-apply) from usage of client-side-apply to server-side-apply:
+
+> Client-side apply users who manage a resource with kubectl apply can start using server-side apply with the following flag.
+>
+>     kubectl apply --server-side [--dry-run=server]
+
+
+Unfortunately it is not so simple. By following these instructions you can be
+left in a situation where after migrating to server-side-apply, certain fields
+become un-deletable from the object without a manual update repairing the managed fields. 
+Here is a simple reproducing case:
+
+Create a configmap in the cluster with client-side apply:
+```shell
+cat <<EOF | kubectl apply -f -
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  key: value
+  legacy: unused
+EOF
+```
+```console
+configmap/test created
+```
+
+Apply the same manifest with `--server-side` flag, as per server-side-apply migration instructions:
+```shell
+cat <<EOF | kubectl apply --server-side -f -
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  key: value
+  legacy: unused
+EOF
+```
+```console
+configmap/test serverside-applied
+```
+
+Remove one of the values from the configmap, we choose `legacy` key, and apply
+again using server-side apply. You would expect that the field is now removed
+from the object, but it is not:
+```shell
+cat <<EOF | kubectl apply --server-side -f -
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: test
+data:
+  key: value
+EOF
+```
+```console
+configmap/test serverside-applied
+```
+
+Check the configmap, expecting the legacy value to be removed, and be surprised that it remains:
+```shell
+kubectl get configmap -o yaml test
+```
+```console
+apiVersion: v1
+kind: ConfigMap
+data:
+  key: value
+  legacy: unused
+metadata:
+  annotations:
+    kubectl.kubernetes.io/last-applied-configuration: |
+      {"apiVersion":"v1","data":{"key":"value"},"kind":"ConfigMap","metadata":{"name":"test","namespace":"default"}}
+  name: test
+  namespace: default
+```
+
+This KEP is intended to fix this problem by providing users with a path to completely
+migrate their managed fields to be used with server-side-apply.
+
+### Goals
+
+This bug prevents users of server-side-apply, which is in GA, from deleting any fields of their objects unless they revert back to using client-side-apply. It especially poses huge
```

Comment:
Above you say that it can only be fixed by manually updating managedFields, and I think that's the correct statement. CSA isn't aware of the managedFields, and as you point out elsewhere, its annotation mirrors SSA's fields, not CSA's, for the duration of SSA being used.

Reply:
corrected, thanks!